src/hotspot/share/runtime/handshake.hpp

```diff
@@ -39,74 +40,75 @@ class JavaThread;
-    return _name;
-  }
+  virtual ~HandshakeClosure() {}
+  const char* name() const    { return _name; }
+  virtual bool is_asynch()    { return false; };
   virtual void do_thread(Thread* thread) = 0;
 };
 
+class AsyncHandshakeClosure : public HandshakeClosure {
+ public:
+   AsyncHandshakeClosure(const char* name) : HandshakeClosure(name) {}
+   virtual ~AsyncHandshakeClosure() {}
+   virtual bool is_asynch()          { return true; }
+};
+
 class Handshake : public AllStatic {
  public:
   // Execution of handshake operation
-  static void execute(HandshakeClosure* hs_cl);
-  static bool execute(HandshakeClosure* hs_cl, JavaThread* target);
-  static bool execute_direct(HandshakeClosure* hs_cl, JavaThread* target);
+  static void execute(HandshakeClosure*       hs_cl);
+  static void execute(HandshakeClosure*       hs_cl, JavaThread* target);
+  static void execute(AsyncHandshakeClosure* hs_cl, JavaThread* target);
 };
 
 // The HandshakeState keeps track of an ongoing handshake for this JavaThread.
-// VMThread/Handshaker and JavaThread are serialized with semaphore _processing_sem
-// making sure the operation is only done by either VMThread/Handshaker on behalf
-// of the JavaThread or by the target JavaThread itself.
+// VMThread/Handshaker and JavaThread are serialized with _lock making sure the
+// operation is only done by either VMThread/Handshaker on behalf of the
+// JavaThread or by the target JavaThread itself.
 class HandshakeState {
   JavaThread* _handshakee;
-  HandshakeOperation* volatile _operation;
-  HandshakeOperation* volatile _operation_direct;
-
-  Semaphore _handshake_turn_sem;  // Used to serialize direct handshakes for this JavaThread.
-  Semaphore _processing_sem;
-  bool _thread_in_process_handshake;
+  FilterQueue<HandshakeOperation*> _queue;
+  Mutex   _lock;
+  Thread* _active_handshaker;
 
-  bool claim_handshake(bool is_direct);
+  bool claim_handshake();
   bool possibly_can_process_handshake();
   bool can_process_handshake();
-  void clear_handshake(bool is_direct);
-
   void process_self_inner();
 
-public:
-  HandshakeState();
+ public:
+  HandshakeState(JavaThread* thread);
 
-  void set_handshakee(JavaThread* thread) { _handshakee = thread; }
+  void add_operation(HandshakeOperation* op);
+  HandshakeOperation* pop_for_self();
+  HandshakeOperation* pop_for_processor();
 
-  void set_operation(HandshakeOperation* op);
-  bool has_operation() const { return _operation != NULL || _operation_direct != NULL; }
-  bool has_specific_operation(bool is_direct) const {
-    return is_direct ? _operation_direct != NULL : _operation != NULL;
+  bool has_operation_for_processor();
+  bool has_operation() {
+    return !_queue.is_empty();
   }
-
-  void process_by_self() {
-    if (!_thread_in_process_handshake) {
-      AutoModifyRestore<bool> temporarily(_thread_in_process_handshake, true);
-      process_self_inner();
-    }
+  bool block_for_operation() {
+    return !_queue.is_empty() || _lock.is_locked();
```

Comment:
A JavaThread is blocked.
A second thread have just executed a handshake operation for this JavaThread and are on line:
https://github.com/openjdk/jdk/blob/cd784a751a3153939b9284898f370160124ca610/src/hotspot/share/runtime/handshake.cpp#L510
And the queue is empty.

The JavaThread wakes up and changes state from blocked to blocked_trans.
It now checks if it's allowed to complete the transition to e.g. vm.

If a third thread adds to queue before the second thread leaves the loop it's operation can be executed.
But the JavaThread could see the queue as empty. (racey as you say)

The executor takes lock and then checks if the JavaThread is safe for processing.
The JavaThread becomes unsafe and then check if lock is locked.
If the lock is locked we must take slow path to avoid this.

We should also take slow path if there is something on queue to processes.
We are unsafe when we check queue and lock is not held, if we 'miss' that anything is on queue, it's fine.
Since any other thread cannot have seen us as safe and seen the item on queue. (since lock is not held)
Thus not allowed to process the operation.